File: .github/workflows/update_theme.py
import json
import os
import subprocess


def get_last_modified_date(directory):
    result = subprocess.run(
        ["stat", "-c", "%y", directory], capture_output=True, text=True
    )
    return result.stdout.split(" ")[0]


def update_theme_json(file_path):
    try:
        with open(file_path, "r") as f:
            content_before = f.read()
            f.seek(0)
            data = json.load(f)

        directory = os.path.dirname(file_path)
        last_modified = get_last_modified_date(directory)
        data["updatedAt"] = last_modified

        with open(file_path, "w") as f:
            json.dump(data, f, indent=2)
            f.truncate()
            content_after = f.read()

        if content_before == content_after:
            print("::set-output name=updated::false")
        else:
            print("::set-output name=updated::true")

        print(f"Updated {file_path} with date {last_modified}")

    except Exception as e:
        print(f"Error processing {file_path}: {e}")
        print("::set-output name=updated::false")


files = subprocess.run(
    ["find", ".", "-name", "theme.json", "-print0"], capture_output=True, text=True
).stdout.split("\0")

for file in filter(None, files):
    update_theme_json(file)

File: .github/scripts/update-theme.js
import { $ } from "bun";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT_DIR = path.resolve(__dirname, "../../");

/**
 * Updates updatedAt field in all theme.json files based on git history
 */
async function updateThemeFiles() {
  let updatedAny = false;

  // Get all theme.json files using Bun's glob
  const files = await $`find . -name "theme.json" -print0`.cwd(ROOT_DIR).quiet().text();
  const fileList = files.split("\0").filter(Boolean);

  for (const file of fileList) {
    const filePath = path.join(ROOT_DIR, file);
    const dirPath = path.dirname(filePath);

    try {
      // Get git modified date for directory
      const relativeDir = path.relative(ROOT_DIR, dirPath);
      const gitDate = await $`git log -1 --format=%ad --date=short -- ${relativeDir}`
        .cwd(ROOT_DIR)
        .quiet()
        .text();
      const trimmedGitDate = gitDate.trim();

      if (!trimmedGitDate) {
        console.log(`No git history for directory ${dirPath}`);
        continue;
      }

      // Read and parse theme.json
      const originalData = JSON.parse(await Bun.file(filePath).text());

      if (originalData.updatedAt === trimmedGitDate) {
        console.log(`Unchanged ${file} — updatedAt already set to ${trimmedGitDate}`);
        continue;
      }

      const originalJson = await Bun.file(filePath).text();

      // Only change the updatedAt field, preserve original formatting for rest of the file
      const lines = originalJson.split("\n");
      const result = lines
        .map((line) => {
          if (line.trim().startsWith('"updatedAt"')) {
            return '  "updatedAt": "' + trimmedGitDate + '",';
          }
          return line;
        })
        .join("\n");

      await Bun.write(filePath, result);

      console.log(`Updated ${file} with date ${trimmedGitDate}`);
      updatedAny = true;
    } catch (e) {
      console.error(`Error processing ${file}:`, e.message);
    }
  }

  return updatedAny;
}

// Run the function
if (import.meta.url === `file://${process.argv[1]}`) {
  const updated = await updateThemeFiles();
  console.log(`Theme files updated: ${updated}`);

  // Set output for GitHub Actions
  if (process.env.GITHUB_OUTPUT) {
    await Bun.write(process.env.GITHUB_OUTPUT, `updated=${updated}\n`, { append: true });
  }
}

export { updateThemeFiles };
